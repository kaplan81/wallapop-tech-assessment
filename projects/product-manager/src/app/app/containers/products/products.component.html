<ul class="products">
  <li *ngFor="let product of products$ | async">{{ product.title }}</li>
</ul>
