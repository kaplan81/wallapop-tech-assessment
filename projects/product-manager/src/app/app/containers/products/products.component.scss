.products {
  &__search {
    display: flex;
    justify-content: center;
    &__sort {
      margin: 33px;
    }
  }
  &__list {
    &__item {
      display: flex;
      justify-content: center;
      mng-product {
        width: 960px;
        margin-bottom: 20px;
      }
    }
  }
}
