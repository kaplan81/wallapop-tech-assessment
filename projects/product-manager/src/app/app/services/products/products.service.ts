import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { delay, map, Observable, tap } from 'rxjs';
import { ProductItem, ProductSearchData } from '../../models/product.model';
import { ProductsStateService } from '../products-state/products-state.service';

@Injectable({
  providedIn: 'root',
})
export class ProductsService {
  constructor(private http: HttpClient, private productsStateService: ProductsStateService) {}

  getProducts(): Observable<ProductItem[]> {
    const url = 'https://frontend-tech-test-data.s3-eu-west-1.amazonaws.com/items.json';
    return this.http
      .get<ProductSearchData>(url)
      .pipe(map((data: ProductSearchData) => data.items))
      .pipe(
        delay(2000),
        tap((entities: ProductItem[]) =>
          this.productsStateService.updateState({
            ...this.productsStateService.state,
            entities,
            loaded: true,
            loading: false,
          }),
        ),
      );
  }
}
