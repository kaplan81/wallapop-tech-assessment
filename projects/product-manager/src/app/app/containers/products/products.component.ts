import { AsyncPipe, NgFor, NgIf } from '@angular/common';
import { ChangeDetectionStrategy, Component, inject, OnDestroy, OnInit } from '@angular/core';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { EMPTY, filter, Observable, shareReplay, Subject, switchMap, takeUntil } from 'rxjs';
import { ProductComponent } from '../../components/product/product.component';
import { ProductItem } from '../../models/product.model';
import { ProductsStateService } from '../../services/products-state/products-state.service';
import { ProductsService } from '../../services/products/products.service';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [NgIf, NgFor, AsyncPipe, ProductComponent, MatProgressSpinnerModule],
  selector: 'mng-products',
  standalone: true,
  styleUrls: ['./products.component.scss'],
  templateUrl: './products.component.html',
})
export class ProductsComponent implements OnInit, OnDestroy {
  #destroyed$ = new Subject<void>();
  loading$: Observable<boolean>;
  #productsService = inject(ProductsService);
  #productsStateService = inject(ProductsStateService);
  products$: Observable<ProductItem[]> = this.#productsStateService.getStateProp('entities');

  constructor() {
    this.loading$ = this.#productsStateService
      .getStateProp('loading')
      .pipe(shareReplay({ bufferSize: 1, refCount: true }), takeUntil(this.#destroyed$));
    // Subscribe to loading$ as soon as possible.
    this.loading$.subscribe();
    this.#productsStateService
      .getStateProp('loaded')
      .pipe(
        filter((loaded: boolean) => !loaded),
        takeUntil(this.#destroyed$),
        switchMap((l: boolean) => {
          if (!l && !this.#productsStateService.state.loading) {
            this.#productsStateService.updateStateProp('loading', true);
            return this.#productsService.getProducts();
          }
          return EMPTY;
        }),
      )
      .subscribe();
  }

  ngOnDestroy(): void {
    this.#destroyed$.next();
  }

  ngOnInit(): void {
    this.#productsStateService.updateStateProp('loading', true);
    this.#productsService.getProducts().subscribe();
  }
}
