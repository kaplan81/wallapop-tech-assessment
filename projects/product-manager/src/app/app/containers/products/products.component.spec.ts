import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { ProductsComponent } from './products.component';

describe('ProductsComponent', () => {
  let fixture: ComponentFixture<ProductsComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        ProductsComponent,
        HttpClientTestingModule,
        RouterTestingModule,
        NoopAnimationsModule,
      ],
    }).compileComponents();
    fixture = TestBed.createComponent(ProductsComponent);
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
