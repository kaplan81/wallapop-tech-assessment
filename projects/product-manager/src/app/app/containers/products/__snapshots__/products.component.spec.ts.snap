// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsComponent should match snapshot 1`] = `
<mng-products
  __ngContext__="0"
  isNextDisabled$={[Function AnonymousSubject]}
  isPrevDisabled$={[Function AnonymousSubject]}
  loading$={[Function Observable]}
  products$={[Function Observable]}
  sort$={[Function BehaviorSubject]}
>
  <div
    class="products"
  >
    <div
      class="products__search"
    >
      <mng-search
        class="products__search__search"
      >
        <form
          class="search ng-untouched ng-pristine ng-valid"
          novalidate=""
        >
          <input
            class="search__input ng-untouched ng-pristine ng-valid"
            formcontrolname="searchCriteria"
            placeholder="Any search gets all products"
            type="text"
          />
        </form>
      </mng-search>
      <mng-sort
        class="products__search__sort"
      >
        <div
          class="sort"
        >
          <form
            class="sort__form ng-untouched ng-pristine ng-valid"
            novalidate=""
          >
            <mat-form-field
              appearance="fill"
              class="mat-mdc-form-field mat-mdc-form-field-type-mat-native-select mat-form-field-no-animations mat-form-field-appearance-fill mat-primary ng-untouched ng-pristine ng-valid"
            >
              <div
                class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled"
              >
                <div
                  class="mat-mdc-form-field-focus-overlay"
                />
                <div
                  class="mat-mdc-form-field-flex"
                >
                  <div
                    class="mat-mdc-form-field-infix"
                  >
                    <label
                      class="mdc-floating-label mat-mdc-floating-label mdc-floating-label--float-above"
                      matformfieldfloatinglabel=""
                    >
                      <mat-label>
                        Sort by
                      </mat-label>
                      <span
                        aria-hidden="true"
                        class="mat-mdc-form-field-required-marker mdc-floating-label--required"
                      />
                    </label>
                    <select
                      aria-invalid="false"
                      aria-required="true"
                      class="mat-mdc-input-element ng-untouched ng-pristine ng-valid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
                      formcontrolname="sortCriteria"
                      matnativecontrol=""
                      required=""
                    >
                      <option
                        value="title"
                      >
                         title 
                      </option>
                      <option
                        value="description"
                      >
                         description 
                      </option>
                      <option
                        value="price"
                      >
                         price 
                      </option>
                      <option
                        value="email"
                      >
                         email 
                      </option>
                    </select>
                  </div>
                </div>
                <div
                  class="mdc-line-ripple mdc-line-ripple--deactivating"
                  matformfieldlineripple=""
                />
              </div>
              <div
                class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
              >
                <div
                  class="mat-mdc-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
                  style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                >
                  <div
                    class="mat-mdc-form-field-hint-spacer"
                  />
                </div>
              </div>
            </mat-form-field>
          </form>
        </div>
      </mng-sort>
    </div>
    <mat-spinner
      aria-valuemax="100"
      aria-valuemin="0"
      class="mat-mdc-progress-spinner mdc-circular-progress mat-primary _mat-animation-noopable mdc-circular-progress--indeterminate"
      mode="indeterminate"
      role="progressbar"
      style="width: 100px; height: 100px;"
      tabindex="-1"
    >
      <div
        aria-hidden="true"
        class="mdc-circular-progress__determinate-container"
      >
        <svg
          class="mdc-circular-progress__determinate-circle-graphic"
          focusable="false"
          viewBox="0 0 100 100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            class="mdc-circular-progress__determinate-circle"
            cx="50%"
            cy="50%"
            r="45"
            style="stroke-dasharray: 282.7433388230814px; stroke-width: 10%;"
          />
        </svg>
      </div>
      <div
        aria-hidden="true"
        class="mdc-circular-progress__indeterminate-container"
      >
        <div
          class="mdc-circular-progress__spinner-layer"
        >
          <div
            class="mdc-circular-progress__circle-clipper mdc-circular-progress__circle-left"
          >
            <svg
              class="mdc-circular-progress__indeterminate-circle-graphic"
              focusable="false"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="50%"
                cy="50%"
                r="45"
                style="stroke-dasharray: 282.7433388230814px; stroke-dashoffset: 141.3716694115407px; stroke-width: 10%;"
              />
            </svg>
          </div>
          <div
            class="mdc-circular-progress__gap-patch"
          >
            <svg
              class="mdc-circular-progress__indeterminate-circle-graphic"
              focusable="false"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="50%"
                cy="50%"
                r="45"
                style="stroke-dasharray: 282.7433388230814px; stroke-dashoffset: 141.3716694115407px; stroke-width: 10%;"
              />
            </svg>
          </div>
          <div
            class="mdc-circular-progress__circle-clipper mdc-circular-progress__circle-right"
          >
            <svg
              class="mdc-circular-progress__indeterminate-circle-graphic"
              focusable="false"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="50%"
                cy="50%"
                r="45"
                style="stroke-dasharray: 282.7433388230814px; stroke-dashoffset: 141.3716694115407px; stroke-width: 10%;"
              />
            </svg>
          </div>
        </div>
      </div>
    </mat-spinner>
  </div>
</mng-products>
`;
